### PR TITLE
refactor(deps): Rewrite 'depends.ps1'

### LINKED ISSUE #3700: Scoop installs 7-Zip even when 7ZIPEXTRACT_USE_EXTERNAL is true
If `Expand-7ZipArchive` is detected in the installer script, 7-Zip is installed even when `7ZIPEXTRACT_USE_EXTERNAL` is set to `true`.

### PR DESCRIPTION
**Pull request by niheaven**
#### Description

Refactor Scoop's dependencies resolving functions, and fix some dependencies error

#### Motivation and Context

- Closes #3700

- Relates to #3502

Commits are organized as following:

- Remove unused code, i.e., `extract_7zip()`, `extract_msi()`, `extract_zip()`, `unpack_inno()` and zip extraction code that before PowerShell 5
- Simplify `Test-xxxRequirement()` that only one type of param is allowed and add tests
- Refactor `script_deps()` and `install_deps()` to `Get-InstallationHelper()` and add tests (fix #3700 in this step)
- Remove naïve function `runtime_deps()` (wrapper of `$manifest.depends`)
- Refactor `dep_resolve()` to `Get-Dependency()`
- Remove `install_order()` and `deps()` with direct call of `Get-Dependency()`
- Refactor code that introduced in #3685 (move some lines of code to make them more readable)
- Fix a small bug related to #3281

#### How Has This Been Tested?

```powershell
# Use LessMSI
scoop config MSIEXTRACT_USE_LESSMSI true
# Uninstall some apps
scoop uninstall 7zip 7zip-zstd lessmsi dark powertoys yori
# Install 7zip-zstd, this will install lessmsi, 7zip and 7zip-zstd
scoop install 7zip-zstd
# Uninstall
scoop uninstall 7zip lessmsi 7zip-zstd
# Use external 7zip
scoop config 7ZIPEXTRACT_USE_EXTERNAL true
# Install yori and this will give error that no 7z.exe while 7ZIPEXTRACT_USE_EXTERNAL is true
scoop install yori
# Change back
scoop config rm 7ZIPEXTRACT_USE_EXTERNAL
# Install multiple apps and it's okay
scoop install yori powertoys 7zip-zstd
```

#### Remarks

If something (i.e., some config option switch) is missing, please let me know.

Maybe `aria2`'s dependence should be added here too? (By config `aria2-enabled`)

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

**Comment by niheaven** [comment]
If there is no problem with this refactoring, I'll merge it and keep moving on some other features.

File: lib/decompress.ps1 [hunk 1]
```diff
@@ -286,34 +231,7 @@ function Expand-ZipArchive {
...
-            }
-        } catch {
-            abort "Unzip failed: $_"
-        }
-    } else {
-        # Use Expand-Archive to unzip in PowerShell 5+
-        # Compatible with Pscx (https://github.com/Pscx/Pscx)
-        Microsoft.PowerShell.Archive\Expand-Archive -Path $Path -DestinationPath $DestinationPath -Force
-    }
+    Expand-Archive -Path $Path -DestinationPath $DestinationPath -Force
```
Comment (chawyehsu) [inline lib/decompress.ps1:234]: You removed the `Microsoft.PowerShell.Archive\` prefix, which was added to fix the confliction with Pscx ...

Comment (niheaven) [reply lib/decompress.ps1:234]: Pscx change their function name to `Expand-PscxArchive` ^_^

Comment (chawyehsu) [reply lib/decompress.ps1:234]: Keep the "Microsoft.PowerShell.Archive" prefix does not have any cons isn't it?

Comment (niheaven) [reply lib/decompress.ps1:234]: Under this release [4.0.0-beta1 - October 17, 2020](https://github.com/Pscx/Pscx/blob/master/ReleaseNotes.txt)

Comment (niheaven) [reply lib/decompress.ps1:234]: I'm afraid this may break something in *nix environment since the `\`

**Pull request merged**